toolbox/deps_update/main.go

```diff
@@ -0,0 +1,259 @@
+	owner      = flag.String("owner", "istio", "Github Owner or org")
+	tokenFile  = flag.String("token_file", "", "File containing Github API Access Token")
+	baseBranch = flag.String("base_branch", "master", "Branch from which the deps update commit is based")
+	githubClnt *githubClient
+)
+
+const (
+	istioDepsFile = "istio.deps"
+)
+
+// Update the commit SHA reference in a given line from dependency file
+// to the latest stable version
+// Returns the updated line
+func replaceCommit(line string, dep dependency) string {
+	idx := strings.Index(line, "\"")
+	return line[:idx] + "\"" + dep.LastStableSHA + "\","
+}
+
+// Generates an MD5 digest of the version set of the repo dependencies
+// useful in avoiding making duplicate branches of the same code change
+// Also updates dependency objects deserialized from istio.deps
+func fingerPrintAndUpdateDepSHA(repo string, deps *[]dependency) (string, error) {
+	digest, err := githubClnt.getHeadCommitSHA(repo, *baseBranch)
+	if err != nil {
+		return "", err
+	}
+	digest += *baseBranch
+	for i, dep := range *deps {
+		commitSHA, err := githubClnt.getHeadCommitSHA(dep.RepoName, dep.ProdBranch)
+		if err != nil {
+			return "", err
+		}
+		digest += commitSHA
+		(*deps)[i].LastStableSHA = commitSHA
+	}
+	return util.GetMD5Hash(digest), nil
+}
+
+// Read the exported environment variable named :key in istio.VERSION
+// and return its value
+func istioVersionGet(key string) (string, error) {
+	// exec always forks a child process to execute the given command, which means
+	// istio.VERSION is sourced to a forked bash process and all exported values
+	// are gone once the child process exits.
+	// The following is a workaround to pass value back to the parent process,
+	// which is whoever executing this go script
+	cmd := fmt.Sprintf("source istio.VERSION ; echo $%s", key)
+	value, err := exec.Command("bash", "-c", cmd).CombinedOutput()
+	return strings.TrimSpace(string(value)), err
+}
+
+// Update the commit SHA reference in the dependency file of dep
+func updateDepFile(dep dependency) error {
+	input, err := ioutil.ReadFile(dep.File)
+	if err != nil {
+		return err
+	}
+	lines := strings.Split(string(input), "\n")
+	for i, line := range lines {
+		if strings.Contains(line, dep.Name+" = ") {
+			lines[i] = replaceCommit(line, dep)
+		}
+	}
+	output := strings.Join(lines, "\n")
+	return ioutil.WriteFile(dep.File, []byte(output), 0600)
+}
+
+// Assumes at the root of istio directory
+// Runs the updateVersion.sh script
+func updateIstioDeps(deps *[]dependency) error {
+	getSHAByName := func(name string) string {
+		for _, d := range *deps {
+			if d.RepoName == name {
+				return d.LastStableSHA
+			}
+		}
+		return ""
+	}
+	pilotSHA := getSHAByName("pilot")
+	mixerSHA := getSHAByName("mixer")
+	AuthSHA := getSHAByName("auth")
+	if pilotSHA == "" || mixerSHA == "" || AuthSHA == "" {
+		return fmt.Errorf("incomplete dependencies in %s", istioDepsFile)
+	}
+	caHub, err := istioVersionGet("CA_HUB")
```

Comment:
I like your idea of updating istio.VERSION directly since ./install/updateVersion.sh will pick up the latest changes I made to istio.VERSION. However, there is a minor twist around istioctlURL as you will see from my comment in code